packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -1933,12 +2011,34 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
 
   @override
   void didChangeNext(Route<dynamic>? nextRoute) {
+    // In some cases, we need to get the delegated transition from a route before
+    // it is built, but some routes need to derive a delegated transition from
+    // the context. So if the next route does not have a `BuildContext` or a
+    // delegatedTransition, call nextRoute.setDelegatedTransition using the
+    // current `BuildContext`.
+    if (nextRoute is ModalRoute<T> && nextRoute.delegatedTransition == null && nextRoute.subtreeContext == null && this.subtreeContext != null) {
+      nextRoute.setDelegatedTransition(this.subtreeContext!);
+    }
+    if (nextRoute is ModalRoute<T> && canTransitionTo(nextRoute) && nextRoute.delegatedTransition.runtimeType != this.delegatedTransition.runtimeType) {
+      if (receivedTransition.runtimeType != nextRoute.delegatedTransition.runtimeType) {
+        receivedTransition = nextRoute.delegatedTransition;
+      }
+    } else {
+      receivedTransition = null;
+    }
     super.didChangeNext(nextRoute);
     changedInternalState();
   }
 
   @override
   void didPopNext(Route<dynamic> nextRoute) {
+    if (nextRoute is ModalRoute<T> && canTransitionTo(nextRoute) && nextRoute.delegatedTransition.runtimeType != this.delegatedTransition.runtimeType) {
```

Comment:
why do we check for `nextRoute.delegatedTransition.runtimeType != this.delegatedTransition.runtimeType`

Reply:
We do not want to override the transitions if the next route has the same transition type. This suppresses possible rebuilds, and gives the pages types that are transitioning to its own type control of that flow, which they have the context to control any adjustments that need to be made.

Reply:
In that case this should be defined in the DelegatedTransition class something like

`DelegateTransitiion.canOverTake`

and in here instead of comparing the runtimeType, you can call that method, it will be up to the subclass to decide whether they want to replace or not

Reply:
The current idea is that if they do want to override a transition from the same route type, then they would define another DelegateTransition subclass. We thought that if a route has multiple possible transitions, then they would either be very different and would want multiple subclasses anyway, or they would vary slightly but it would be unlikely that they would have enough variations to require making more than a couple subclasses. And DelegatedTransition is a fairly lightweight class so it's not difficult to do. The subclasses also have controls they can put in using `canTransitionTo`.

We could do something like `DelegateTransition.canOverTake`, but it would likely have a default similar to checking the type.

Reply:
> The current idea is that if they do want to override a transition from the same route type, then they would define another DelegateTransition subclass

I am not sure we can assume this. The current implementation is surprising to me. If the idea behind using the runtimeType is to not overtake the transition of same time, can we make either override the == method of the subclass or make all the delegateTransition to use the same instance like some thing define statically in MaterialPageRoute or transitiontheme

Reply:
> can we make either override the == method of the subclass

We did try this in an earlier version for DelegatedTransition. The comparison was against a `name` key property added to DelegatedTransition, which we went back on out of concerns for collision. Comparing the builder for equality was also unreliable, even when they were pointing at the same static function on the transition class, especially on pop. I'm not sure how to override the == method in a way other than checking the runtime type.

> make all the delegateTransition to use the same instance like some thing define statically in MaterialPageRoute or transitiontheme

For the zoom route in particular that would be odd because of the snapshot arguments. We'd have to add 4 different static instances for that, all of which would not want to override each other. 

If we tried to use the transition theme that would create issues too. Cupertino has no transition theme, and for developers making a custom route with a custom transition using this logic, we'd have to direct them towards making a static function that they would add into a theme?

Reply:
actually, is this the cases for all DelegateTransition subclass? or this doesn't matter for most subclasses except some.

Reply:
> actually, is this the cases for all DelegateTransition subclass? or this doesn't matter for most subclasses except some.

Of the transitions we have, only the Zoom transition can have differences between two transitions that we do not want to override the previous page with, to let each page determine it's own snapshotting logic. The rest of the transitions are the same as each other, and we still do not want to override the previous transition with them.

Modal bottom sheet is going to have two different DelegateTransitions that will want to override each other, so they would be different subclasses.